File: Algorithm_80_Remove Duplicates from Sorted Array II.py
# 思路：用标志位repeated记录当前元素是否发生过重复
# 如果nums[i] == nums[i - 1]而repeated == False，表明是第一次重复，可以允许
# 把nums[i]放到用于记录当前长度的变量newLen位置上
# 若repeated == True，则无视该元素，直接进入下一层循环
# 若nums[i] != nums[i - 1]，表明nums[i]是第一次出现，直接放到newLen位置上
# 并把repeated标志置为False，表明是第一次出现

class Solution:
    def removeDuplicates(self, nums: List[int]) -> int:
        if len(nums) <= 2:
            return len(nums)
        newLen = 1
        repeated = False
        for i in range(1, len(nums)):
            if nums[i] == nums[i - 1]:
                if not repeated: # allow repeat for 1 time
                    repeated = True
                    nums[newLen] = nums[i]
                    newLen += 1
            else:
                repeated = False
                nums[newLen] = nums[i]
                newLen += 1
        return newLen

File: Algorithm_80_Remove Duplicates from Sorted Array II.java
// 思路：双指针。posToInsert指向下一个要插入的index
// 对于当前的元素nums[cur]，如果是第一次出现，则插入到nums[posToInsert]上，并且posToInsert++
// 并且置repeated = false，表示当前元素还未重复过
// 如果nums[cur] == nums[cur - 1]，说明当前元素并非第一次出现
// 此时，如果repeated == false，说明之前还未重复使用过这个元素
// 因此依然可以把它插入到nums[posToInsert]位置上，并且posToInsert++
// 否则，不做任何处理

class Solution {
    public int removeDuplicates(int[] nums) {
        int posToInsert = 1, cur = 1;
        boolean repeated = false;
        for (cur = 1; cur < nums.length; cur++) {
            if (nums[cur] == nums[cur - 1]) {
                if (!repeated) {
                    nums[posToInsert++] = nums[cur];
                    repeated = true;
                }
            } else {
                nums[posToInsert++] = nums[cur];
                repeated = false;
            }
        }
        return posToInsert;
    }
}